pyo3-macros-backend/src/pyclass.rs

```diff
@@ -512,6 +531,56 @@ fn impl_enum(
     impl_enum_class(enum_, args, doc, methods_type, krate)
 }
 
+fn impl_into_py_enum(cls: &syn::Ident, variants: &[PyClassEnumVariant<'_>]) -> TokenStream {
+    // Assuming all variants are unit variants because they are the only type we support.
+    match variants.len() {
+        1 => {
+            let variant_name = variants.first().unwrap().ident;
+            quote! {
+                impl _pyo3::IntoPy<_pyo3::PyObject> for #cls {
+                    fn into_py(self, py: _pyo3::Python) -> _pyo3::PyObject {
+                        static SINGLETON: _pyo3::once_cell::GILOnceCell<_pyo3::Py<#cls>> = _pyo3::once_cell::GILOnceCell::new();
+
+                        let singleton = SINGLETON.get_or_init(py, || {
+                            _pyo3::Py::new(py, #cls::#variant_name).unwrap()
+                        });
+
+                        _pyo3::IntoPy::into_py(::std::clone::Clone::clone(singleton), py)
+                    }
+                }
+            }
+        }
+        _ => {
+            let initialize_values: Vec<TokenStream> = variants
+                .iter()
+                .map(|variant| {
+                    let variant_name = variant.ident;
+                    quote!((#cls::#variant_name as usize, _pyo3::Py::new(py, #cls::#variant_name).unwrap()))
+                })
+                .collect();
+
+            quote! {
+                impl _pyo3::IntoPy<_pyo3::PyObject> for #cls {
+                    fn into_py(self, py: _pyo3::Python) -> _pyo3::PyObject {
+                        static SINGLETON_PER_VARIANT: _pyo3::once_cell::GILOnceCell<::std::collections::HashMap<usize, _pyo3::Py<#cls>>> = _pyo3::once_cell::GILOnceCell::new();
```

Comment:
Is there a reason to use a `HashMap` here instead of a fixed-size array?

Reply:
I think `#cls::#variant_name as usize` could be sparse, e.g.

```
enum Foo {
  Bar = 1,
  Baz = 999999,
}
```

so while we could store those as sorted pairs in a `[(usize, Py<#cls>); 2]`, we would need to do a binary search for the variants instead of a hash table look-up as direct access into a `[(usize, Option<Py<#cls>>); 1000000]` seems wasteful.

Personally, I see the simplicity of using the hash table but would probably opt for the sorted array in a follow-up. (But let's whether this lands using the `static`s at all.)

Reply:
The discriminant can indeed be sparse which is also tested by `test_custom_discriminant_comparison_by_identity `. But using binary search in a sorted array is definitely a possible replacement for the hashmap.

Reply:
I think we can know from parsing the enum whether it's dense or sparse?

For the dense case we can just index into an array, no binary search needed, and for the sparse case we can generate a `match` in front to convert into a dense set of indices first.

Reply:
> and for the sparse case we can generate a match in front to convert into a dense set of indices first.

This would really be nice as the compiler is pretty clever in how to handle `match` AFAIK, e.g. producing linear or binary searches depending on cardinality estimates.